codegen/smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/discovery/EndpointDiscovererGeneratorTest.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.endpoints.discovery
+
+import org.junit.jupiter.api.Test
+import software.amazon.smithy.build.MockManifest
+import software.amazon.smithy.kotlin.codegen.test.*
+
+class EndpointDiscovererGeneratorTest {
+    @Test
+    fun testClass() {
+        val actual = render()
+
+        actual.shouldContainOnlyOnceWithDiff(
+            """
+            public class TestEndpointDiscoverer {
+                private val cache = ReadThroughCache<DiscoveryParams, Host>(10.minutes, Clock.System)
+            """.trimIndent(),
+        )
+
+        actual.shouldContainOnlyOnceWithDiff(
+            """
+                }
+                
+                private val discoveryParamsKey = AttributeKey<DiscoveryParams>("DiscoveryParams")
+                private data class DiscoveryParams(private val region: String, private val identity: String)
+            """.trimIndent(),
+        )
+    }
+
+    @Test
+    fun testAsEndpointResolver() {
+        val actual = render()
+
+        actual.shouldContainOnlyOnceWithDiff(
+            """
+                internal suspend fun asEndpointResolver(client: TestClient, delegate: EndpointResolverAdapter) = EndpointResolver { request ->
```

Comment:
**correctness**: Why does the conversion `asEndpointResolver` need to be `suspend`?

Reply:
Good catch, it does not need to be.